### DISCUSSION TITLE
Table in a box has narrow columns :-(

### DISCUSSION THREAD

**Discussion by tfcroft4** [beeware/toga#2973] [Debugging help]
I am building a GUI with a table.
If I place a table in a box and add the box to the main window the table columns are compressed. 
If I add a table directly to the main window the table expands as expected.
Perhaps I have missed something.
Is there a container setting needed to allow the table to expand? 
(direction setting does not seem to have any effect)
I am running on Windows 10. beeware installed as per Getting Started tutorial on 19/11/2024

```
main_box = toga.Box(style=Pack(direction=COLUMN))
data_box = toga.Box(style=Pack(direction=COLUMN, padding=5 , background_color=BLUE)) 
data_table = toga.Table(style=Pack(flex=1,padding=5),
        headings=["Name", "Age", "Planet"],
        #accessors={"Name", 'character'},
        data=[
        {"name": "Arthur Dent", "age": 42, "planet": "Earth"},
        {"name": "Ford Prefect", "age": 37, "planet": "Betelgeuse Five"},
        {"name": "Tricia McMillan", "age": 38, "planet": "Earth"},
        ]   
        )
data_table_2 = toga.Table(style=Pack(flex=1,padding=5),
        headings=["Planet","Name", "Age",],
        #accessors={"Name", 'character'},
        data=[
        {"name": "Tricia McMillan", "age": 38, "planet": "Earth"},
        {"name": "Ford Prefect", "age": 37, "planet": "Betelgeuse Five"},
        {"name": "Arthur Dent", "age": 42, "planet": "Earth"},
        ]   
        )
data_box.add(data_table)
main_box.add(data_box) #compressed table width
main_box.add(data_table_2) #full table width
```
![Toga Table Test](https://github.com/user-attachments/assets/6ccfc4d2-a105-4b93-bdf7-6fd79d8197ca)

**Comment by freakboy3742** [comment]
That definitely sounds like a bug to me. You definitely haven't missed a "magic setting" to make the window expand; it should be expanding columns to fill (as it does in the second example). I've logged #2974 to track this as a bug.

**Reply by tfcroft4** [reply]
Thanks for the prompt response. 
So far this has been the only issue in building the GUI.
Thanks for your work.
Ted

**Comment by tfcroft4** [comment]
Additionally a table in an opcontainer does not always reflect changes in the underling data ListSource.
I found that a programmatic change to data was not shown until a subsequent action. In my case a row delete forced a data display update.

**Reply by freakboy3742** [reply]
If you believe you've found a bug, then I encourage you to open a bug report with a reproduction case. There's very little we're able to do when all the details we have are "an opcontainer does not always reflect changes".

**Comment by steve64** [comment]
Any timeframe to solve the table bug related to column sizes?
I would like to use Toga in a commercial app, but this bug is preventing this at current stage.

I need to add the the bug is not always shown. I found cases where the column are properly dimensioned
with equal size along all the available horizontal space, even if the table is placed inside a box. 
I still do not know how to precisely trigger this good behavior,
but refilling a table after first time seems to help.

By the way, IMHO, the column sizes should be individually programmable...

**Reply by freakboy3742** [reply]
I've flagged everything I know on the ticket (#2974).

Toga is an open source project. It's offered as is; and there are known bugs. This is one of those bugs.

This issue isn't currently a priority for me personally (if only because I'm currently on leave for the new year break). I can't offer a timeframe for work that isn't currently on my priority list.

If you want to investigate and propose a fix for one of those bugs, I'm happy to review it. If you have a commercial interest, the way to get this onto someone's priority list is to pay them so that your problems are their priority. I may be able to offer those services; if that is of interest to you, [send me an email](mailto:rkeith-magee@anaconda.com) and I we can negotiate terms. However, I'm not the only person that could do this work; again, this is an Open Source project, and anyone can submit a PR.
